### DISCUSSION TITLE
Connect PostgreSQL DB inside Kubernetes Cluster to a Teleport Cluster

### DISCUSSION THREAD

**Discussion by janlauber** [gravitational/teleport#10385] [Q&A] [answered]
Hey folks

I'm wondering if it's possible to add a PostgreSQL Instance running inside a Kubernetes Cluster (which is also joined via the teleport-agent) to the teleport cluster.

I'm trying to map the Documentation on self hosted PostgreSQL to the Kubernetes Bitnami Helm Chart of PostgreSQL.

Does anyone have some suggestions?

Greez,
Jan

**Comment by blushingpenguin** [comment]
@janlauber did you find a way of doing this? I'm looking to do the same thing myself.  Thanks!

**Reply by janlauber** [reply]
There are some configurations in the helm chart of the teleport-agent to connect a postgresql database inside the cluster. It will show up in the webinterface but the connection doesn't work. Idk what I'm doing wrong but I went back to legacy Database servers and trying to connect them to the teleport cluster (also not working rn)

**Reply by jhoelzel** [reply]
Dear @janlauber  have you ever figured this out? Im stuck in the same loop.

**Reply by janlauber** [reply]
@jhoelzel 
This is our documentation to do it:

# Adding a PostgreSQL Database Server
[Database Access with Self-Hosted PostgreSQL | Teleport Docs](https://goteleport.com/docs/database-access/guides/postgres-self-hosted/)  

The Connection to postgreSQL is done using mutual TLS authentication.

This requires:

- Locally installed tctl client and signed in with an admin
- postgreSQL Server with enabled TLS Authentication
- Teleport CA installed on postgreSQL Server
- Valid Server Certificate for postgreSQL Server
- SQL User with Authentication over TLS

Create the required certificates:

```bash
tsh login --proxy=teleport.<cust>.k8s.example.com:443 --auth=local --user=<user>
tctl auth sign --format=db --host=${IP_HOST} --out=server --ttl=86400h
```

> Be sure you set the permissions and owner correct. Have a look at the logfile after starting postgresql: less /var/log/postgresql/postgresql-14-main.log

Copy the certificates to the postgreSQL Server and change permissions:

```bash
scp server.* user@postgreSQL:/etc/postgresql/14/main/conf.d
chmod 600 server.*
chown postgres:postgres server.*
```

Enable TLS on postgreSQL Server:

```bash
nano /etc/postgresql/14/main/conf.d/pg_teleport.conf

ssl = on
ssl_cert_file = '/etc/postgresql/14/main/conf.d/server.crt'
ssl_key_file = '/etc/postgresql/14/main/conf.d/server.key'
ssl_ca_file = '/etc/postgresql/14/main/conf.d/server.cas'

systemctl restart postgresql
journalctl -u postgresql
```

Enable TLS Authentication using certificates:

> Make sure this config is at the top of pg_hba.conf to ensure, it will be picked first
> This config is automatically done, by our ansible postgres role.

```
# Enforce Certificate Authentication when connecting over TLS
hostssl all             all             ::/0                    cert
hostssl all             all             0.0.0.0/0               cert

systemctl restart postgresql
journalctl -u postgresql
```

Create one or more teleport users, that can access the postgreSQL server.

```bash
sudo -u postgres psql

# This role doesn`t need a password as authentication will be done by Certificate
CREATE ROLE teleport_customer_developer WITH LOGIN SUPERUSER; 
```

Now you need to make sure, that the same username is allowed in the teleport role, to access the database:

```yaml
spec:
  allow:
    db_names:
    - postgres-prod # Name of Databases allowed to access
    - '*' # Allow all databases
    db_users:
    - teleport_customer_develeoper # Name of user to access the Database
```

I hope this helps.

**Reply by jhoelzel** [reply]
Thank you very much for this! I think the main magic is simply the certificate authentification =)

**Comment by webvictim** [comment, marked as answer]
Yes, you should be able to do this by adding the `databases:` section to the agent chart config and using a join token with appropriate permissions for `kube,db` services.

If you're having issues, please post details of your configs/logs and I'll try to help you fix it.
